ddtrace/propagation/http.py

```diff
@@ -483,10 +489,201 @@ def _extract(headers):
+                span_id = _hex_id_to_dd_id(span_id_hex)
+
+                # All 0s are invalid values
+                assert trace_id != 0
+                assert span_id != 0
+
+                trace_flags = _hex_id_to_dd_id(trace_flags_hex)
+                # there's currently only one trace flag, which denotes sampling priority
+                # was set to keep "01" or drop "00"
+                # trace flags is a bit field: https://www.w3.org/TR/trace-context/#trace-flags
+                sampling_priority = int(trace_flags & 0x1)
+
+            else:
+                log.debug("W3C traceparent hex length incorrect: %s", tp)
+                raise ValueError
+
+        except (ValueError, AssertionError):
+            log.exception("received invalid w3c traceparent: %s.", tp)
+            return None
+
+        return tp, trace_id, span_id, sampling_priority
+
+    @staticmethod
+    def _get_tracestate_values(ts):
+        # type: (str) -> Optional[Tuple[Optional[int], Dict[str, str], Optional[str]]]
+
+        # tracestate parsing, example: dd=s:2;o:rum;t.dm:-4;t.usr.id:baz64,congo=t61rcWkgMzE
+        # store ts so we keep other vendor data
+        meta = {_TRACESTATE_KEY: ts}
+
+        try:
+            result = re.search("dd=(.+?),", ts)
+
+            # grab dd value if dd is only list member
+            if not result:
+                result = re.search("dd=(.*)", ts)
+            try:
+                dd = dict(item.split(":") for item in result.group(1).split(";"))  # type: ignore
+            except AttributeError:
+                log.debug(("no dd list member in tracestate from incoming request: %r "), ts)
+                dd = None
+
+            # parse out values
+            if dd:
+                sampling_priority_ts = dd.get("s")
+                if sampling_priority_ts:
+                    sampling_priority_ts = int(sampling_priority_ts)
+                else:
+                    sampling_priority_ts = None
+
+                origin = dd.get("o")
+                # need to convert from t. to _dd.p.
+                other_propagated_tags = {
+                    re.sub("t.", "_dd.p.", k): v
+                    for (k, v) in dd.items()
+                    if (_TraceContext._is_valid_datadog_trace_tag_key(k))
+                }
+                meta.update(other_propagated_tags)
+
+                return sampling_priority_ts, meta, origin
+            else:
+                # even if there's no dd list member, we still need to propagate the other tracestate values
+                return None, meta, ""
+
+        except (TypeError, ValueError, AttributeError):
+            log.debug(("received invalid dd header value in tracestate: %r "), ts)
+            return None
+
+    @staticmethod
+    def _decicide_sampling_priority(sampling_priority_tp, sampling_priority_ts):
+        # type: (int, int) -> int
+        if sampling_priority_tp == 0 and (not sampling_priority_ts or sampling_priority_ts >= 0):
```

Comment:
Can we add a doc string with the expected logic. It will make this these if statements easier to review and maintain.